### ISSUE TITLE
research: prototype an architecture-specific HRM-Text-1B adaptation lane

### ISSUE THREAD

**Issue by rmems** [rmems/agoge-forger#107]
## Goal

Create a separately versioned adaptation lane for `sapientinc/HRM-Text-1B` that respects its recurrent/PrefixLM training semantics instead of treating it as another routine causal-LM QLoRA target.

Parent: #76
Compatibility gate: #104
Shared data/eval contracts: #99, #100

## Research role

HRM-Text is the **high-risk architecture study** in the approved model cut.

Its purpose is to determine whether Agoge can support a recurrent/PrefixLM-style checkpoint with explicit masking/serialization semantics, and only then decide whether a measured engineering-data adaptation is scientifically comparable enough to run.

## Guardrail

A successful generic PEFT smoke is **not** sufficient evidence that the upstream HRM SFT semantics were reproduced.

Before measured training, inspect and document:

- exact Hub revision and upstream code revision
- model class and recurrent slow/fast stack structure
- PrefixLM attention/masking behavior
- instruction vs response loss-masking policy
- upstream data-preparation expectations
- trainable-module inventory
- whether LoRA/QLoRA is technically supported and what it changes
- whether a full-parameter or architecture-native path is required for the intended claim

### Stage A — compatibility

Under #104:

- load checkpoint/tokenizer
- inspect model graph
- run forward/generation smoke
- validate candidate target modules without assuming LLaMA names
- tiny-overfit only as a mechanical test
- save/reload any adapter if used

### Stage B — semantics validation

Build fixtures proving that the Agoge-rendered example, attention mask, position behavior, and loss mask match the declared HRM adaptation contract.

### Stage C — measured adaptation decision

Only after A/B pass, freeze a separate experiment contract. Reuse #99 source-level example identities and #100 logical held-out tasks, but explicitly disclose any HRM-specific rendering/masking differences.

## Required artifacts

```text
reports/hrm-text-adaptation/<experiment_id>/
  architecture.json
  upstream-contract.md
  masking-fixtures.jsonl
  target-modules.json
  compatibility.json
  experiment-contract.json      # only if Stage C is authorized
  report.md
```

## Acceptance criteria

- [ ] Exact HRM-Text and upstream code revisions are pinned.
- [ ] PrefixLM/recurrent architecture behavior is documented from executable inspection, not assumed from a generic Transformer path.
- [ ] Loss-masking and serialization fixtures prove what tokens contribute to SFT loss.
- [ ] Candidate PEFT targets are architecture-derived and explicitly frozen.
- [ ] Any adapter save/reload smoke is labeled as compatibility evidence only until semantics validation passes.
- [ ] A go/no-go decision is recorded for a measured HRM engineering-data adaptation.
- [ ] Any eventual HRM result is reported as a separately versioned architecture condition, not silently pooled with Granite/Mellum/LFM2.5.

## Non-goals

- claiming reproduction of upstream HRM training from a generic TRL/PEFT run
- forcing HRM into the same implementation path merely for matrix symmetry
- spending cloud credits before the local semantics/compatibility gates pass

## Planning attribution

Approved model role from the 2026-08-28 Agoge model cut; issue drafted with OpenAI ChatGPT — GPT-5.6 Sol.

**Comment by rmems** [comment]
Closing issue as this repo continues to collect tech debt
